src/main/java/com/redhat/red/build/finder/BuildFinder.java

```diff
@@ -117,31 +136,220 @@ public BuildFinder(ClientSession session, BuildConfig config) {
+
+                if (request != null) {
+                    LOGGER.debug("Got task request for build id {}: {}", buildInfo.getId(), request);
+
+                    KojiTaskRequest taskRequest = new KojiTaskRequest(request);
+
+                    build.setTaskRequest(taskRequest);
+                } else {
+                    LOGGER.debug("Null task request for build id {} with task id {} and checksum {}", red(buildInfo.getId()), red(taskInfo.getTaskId()), red(checksum));
+                }
+            } else {
+                LOGGER.debug("Task info not found for build id {}", red(buildInfo.getId()));
+            }
+        } else {
+            LOGGER.debug("Found import for build id {} with checksum {} and files {}", red(buildInfo.getId()), red(checksum), red(filenames));
+        }
+
+        addArchiveToBuild(build, archive, filenames);
+
+        List<String> buildTypes = null;
+
+        if (buildInfo.getTypeNames() != null) {
+            buildTypes = new ArrayList<>();
+            buildTypes.addAll(buildInfo.getTypeNames());
+        }
+
+        build.setRemoteArchives(allArchives);
+        build.setTags(tags);
+        build.setTypes(buildTypes);
+
+        return build;
+    }
+
+    private void addArchiveWithoutBuild(String checksum, Collection<String> files) {
+        KojiBuild buildZero = builds.get(0);
+        Optional<KojiLocalArchive> matchingArchive = buildZero.getArchives().stream().filter(a -> a.getArchive().getChecksum().equals(checksum)).findFirst();
+        List<String> filenames;
+
+        if (matchingArchive.isPresent()) {
+            KojiLocalArchive existingArchive = matchingArchive.get();
+            filenames = existingArchive.getFiles();
+
+            LOGGER.debug("Adding not-found checksum {} to existing archive {}", existingArchive.getArchive().getArchiveId());
+
+            filenames.addAll(files);
+        } else {
+            filenames = new ArrayList<>(files);
+
+            KojiArchiveInfo tmpArchive = new KojiArchiveInfo();
+
+            tmpArchive.setBuildId(0);
+            tmpArchive.setFilename("not found");
+            tmpArchive.setChecksum(checksum);
+            tmpArchive.setChecksumType(config.getChecksumType());
+
+            tmpArchive.setArchiveId(-1 * (buildZero.getArchives().size() + 1));
+
+            LOGGER.debug("Adding not-found checksum {} to new archive {}", tmpArchive.getArchiveId());
+
+            buildZero.getArchives().add(new KojiLocalArchive(tmpArchive, filenames));
+        }
+
+        filenames.sort(Comparator.naturalOrder());
+    }
+
+    private void addArchiveToBuild(KojiBuild build, KojiArchiveInfo archive, Collection<String> files) {
+        List<String> filenames = new ArrayList<>(files);
+
+        LOGGER.debug("Adding all filenames with archive id {} with files {}", archive.getArchiveId(), filenames);
+
+        if (build.getArchives() == null) {
+            build.setArchives(new ArrayList<>());
+        }
```

Comment:
It's only initializing this for build 0. Would it make sense to initialize a new `ArrayList` in the `KojiBuild` constructors?

Reply:
Note that there are a couple other places in the code where a list is checked for null.

Reply:
+1 on just initialise the ArrayList I think, as it would avoid the null checks

Reply:
Yes, the lazy object creation just seems to be making the code more complicated. Since we intend to modify the list itself and don't provide methods to do so, we can't just return `Collections.emptyList()`.